README.md
OpenST.js
============

OpenST is a framework for building token economies. OpenST supports interactions with openst-contracts.
It requires [openst-contracts](https://github.com/openst/openst-contracts) for ABIs and BINs of contracts.

![Build master](https://img.shields.io/travis/openst/openst.js/master.svg?label=build%20master&style=flat)
![Build develop](https://img.shields.io/travis/openst/openst.js/develop.svg?label=build%20develop&style=flat)
![npm version](https://img.shields.io/npm/v/@openst/openst.js.svg?style=flat)
![Discuss on Discourse](https://img.shields.io/discourse/https/discuss.openst.org/topics.svg?style=flat)
![Chat on Gitter](https://img.shields.io/gitter/room/openst/SimpleToken.svg?style=flat)

##  Setup
This library assumes that nodejs and geth are installed and running. To install OpenST.js in your project using npm:

```bash
$ npm install @openst/openst.js --save
```

The code works for Ethereum Byzantium and Petersburg.

## Creating an OpenST object

```js
// Creating OpenST.js object
const OpenST = require('@openst/openst.js');

```

## Deploying contracts

## Constants
Before deploying contracts, please set some constants to funded addresses that you control.

```js

// Initialize web3 object using the geth endpoint
const Web3 = require('web3');
const web3Provider = new Web3('http://127.0.0.1:8545');

// deployer address
const deployerAddress = '0xaabb1122....................';

// Worker address
const worker = '0xaabb1122....................';

// Relayer address
const relayer = '0xaabb1122....................';

// It's needed for unlock account before doing transaction.
const passphrase = 'some passphrase.....';

// Other constants
const gasPrice = '0x12A05F200';
const gas = '7500000';

```

### Deploy Organization contract

An Organization contract serves as an on-chain access control mechanism by assigning roles to a set of addresses.

```js
const { Organization } = OpenST.ContractInteract;
const organizationOwner = '0xaabb1122....................';
const admin = '0xaabb1122....................';
const orgConfig = {
  deployer: deployerAddress,
  owner: organizationOwner,
  admin: admin,
  workers: [worker],
  workerExpirationHeight: '200000000' // This is the block height for when the the worker is set to expire.
};
const txOptions = {
  gasPrice: '0x3B9ACA00',
  from : deployerAddress,
  gas: '7500000', // This is an optional parameter, if not passed gas will be estimated.
};
let organizationContractInstance;
let organizationAddress;
Organization.setup(web3Provider, orgConfig, txOptions).then(function(instance){
  organizationContractInstance = instance;
  organizationAddress = organizationContractInstance.address;
});

```

### Deploy EIP20Token contract

To perform economy setup, you will need an EIP20Token. You can either use an existing EIP20Token contract or deploy a new one.

### Deploy TokenRules contract

One TokenRules contract is deployed per Organization. Only the Organization whitelisted workers can register rule contracts in the TokenRules contract.

```js
txOptions = {
  from: deployerAddress,
  gasPrice: gasPrice,
  gas: gas
};
const tokenRulesSetupObject = new OpenST.Setup.TokenRules(web3Provider);
let tokenRulesAddress;
tokenRulesSetupObject.deploy(organization, eip20Token, txOptions).then(function(response){
  tokenRulesAddress = response.receipt.contractAddress;
})

```   

### Deploy TokenHolder MasterCopy

```js
txOptions = {
  from: deployerAddress,
  gasPrice: gasPrice,
  gas: gas
};
const userSetup = new OpenST.Setup.User(web3Provider);
let tokenHolderMasterCopyAddress;
userSetup.deployTokenHolderMasterCopy(txOptions).then(function(response){
  tokenHolderMasterCopyAddress = response.receipt.contractAddress;
});

```  

### Deploy Gnosis MasterCopy

```js
txOptions = {
  from: deployerAddress,
  gasPrice: gasPrice,
  gas: gas
};
const userSetup = new OpenST.Setup.User(web3Provider);
let gnosisMasterCopyAddress;
userSetup.deployMultiSigMasterCopy(txOptions).then(function(response){
  gnosisMasterCopyAddress = response.receipt.contractAddress;
});

```  

### Deploy DelayedRecoveryModule MasterCopy

```js
txOptions = {
  from: deployerAddress,
  gasPrice: gasPrice,
  gas: gas
};
const userSetup = new OpenST.Setup.User(web3Provider);
let delayedRecoveryModuleMasterCopyAddress;
userSetup.deployDelayedRecoveryModuleMasterCopy(txOptions).then(function(response){
  delayedRecoveryModuleMasterCopyAddress = response.receipt.contractAddress;
});

```  

### Deploy CreateAndAddModules Contract

```js
txOptions = {
  from: deployerAddress,
  gasPrice: gasPrice,
  gas: gas
};
const userSetup = new OpenST.Setup.User(web3Provider);
let createAndAddModulesAddress;
userSetup.deployCreateAndAddModules(txOptions).then(function(response){
  createAndAddModulesAddress = response.receipt.contractAddress;
});

```  

### Deploy UserWalletFactory Contract

```js
txOptions = {
  from: deployerAddress,
  gasPrice: gasPrice,
  gas: gas
};
const userSetup = new OpenST.Setup.User(web3Provider);
let userWalletFactoryAddress;
userSetup.deployUserWalletFactory(txOptions).then(function(response){
  userWalletFactoryAddress = response.receipt.contractAddress;
});

```

### Deploy ProxyFactory Contract

```js
txOptions = {
  from: deployerAddress,
  gasPrice: gasPrice,
  gas: gas
};
const userSetup = new OpenST.Setup.User(web3Provider);
let proxyFactoryAddress;
userSetup.deployProxyFactory(txOptions).then(function(response){
  proxyFactoryAddress = response.receipt.contractAddress;
});

``` 

### Deploy PricerRule Contract

```js
txOptions = {
  from: deployerAddress,
  gasPrice: gasPrice,
  gas: gas
};
const rulesSetup = new OpenST.Setup.Rules(web3Provider);
const baseCurrencyCode = 'OST';
const conversionRate = 10;
const conversionRateDecimals = 5;
const requiredPriceOracleDecimals = 18;
let pricerRuleAddress;
rulesSetup.deployPricerRule(baseCurrencyCode, conversionRate, conversionRateDecimals, requiredPriceOracleDecimals, txOptions).then(function(response){
  pricerRuleAddress = response.receipt.contractAddress;
})

``` 

## Registration of Rule to TokenRules

```js
txOptions = {
  from: worker,
  gasPrice: gasPrice,
  gas: gas
};
const pricerRule = 'PricerRule';
const tokenRules = new OpenST.Helpers.TokenRules(tokenRulesAddress, web3Provider);
const pricerRuleAbi = abiBinProvider.getABI(pricerRule);
tokenRules.registerRule(pricerRule, pricerRuleAddress, pricerRuleAbi.toString(), txOptions).then(function(response){
  console.log("Successfully registered PricerRule");
})

``` 

## Creating a User Wallet

```js
txOptions = {
  from: deployerAddress,
  gasPrice: gasPrice,
  gas: gas
};
const User = new OpenST.Helpers.User(
  tokenHolderMasterCopyAddress, 
  gnosisMasterCopyAddress, 
  delayedRecoveryModuleMasterCopyAddress, 
  createAndAddModulesAddress, 
  eip20Token, 
  tokenRulesAddress,
  userWalletFactoryAddress, 
  proxyFactoryAddress, 
  web3Provider
);
const owner = '0xaabb1122....................';
const ephemeralKey = '0xaabb1122....................';
const sessionKeySpendingLimit = '1000000';
const sessionKeyExpirationHeight = '100000000000'; 
const threshold = 1;
const recoveryOwnerAddress = '0xaabb1122....................';
const recoveryControllerAddress = '0xaabb1122....................';
const recoveryBlockDelay = 10;
let gnosisSafeProxy;
let userTokenHolderProxy;
User.createUserWallet(
  [owner], 
  threshold, 
  recoveryOwnerAddress, 
  recoveryControllerAddress, 
  recoveryBlockDelay, 
  [ephemeralKey], 
  [sessionKeySpendingLimit], 
  [sessionKeyExpirationHeight], 
  txOptions
).then(function(response){
  const returnValues = response.events.UserWalletCreated.returnValues;
  const userWalletEvent = JSON.parse(JSON.stringify(returnValues));
  gnosisSafeProxy = userWalletEvent._gnosisSafeProxy;
  userTokenHolderProxy = userWalletEvent._tokenHolderProxy;
});

``` 

## Creating a Company Wallet

```js
txOptions = {
  from: deployerAddress,
  gasPrice: gasPrice,
  gas: gas
};
const User = new OpenST.Helpers.User(
  tokenHolderMasterCopyAddress,
  null, // gnosis safe master copy address
  null, // delayed recovery module master copy address
  null, // create and add modules contract address
  eip20Token,
  tokenRulesAddress,
  userWalletFactoryAddress,
  proxyFactoryAddress,
  web3Provider
);
const owner = '0xaabb1122....................';
const ephemeralKey = '0xaabb1122....................';
const sessionKeySpendingLimit = '1000000';
const sessionKeyExpirationHeight = '100000000000'; 
let companyTokenHolderProxy;
User.createCompanyWallet(
  owner,
  [ephemeralKey],
  [sessionKeySpendingLimit],
  [sessionKeyExpirationHeight],
  txOptions
).then(function(response){
  const returnValues = response.events.ProxyCreated.returnValues;
  const proxyEvent = JSON.parse(JSON.stringify(returnValues));
  companyTokenHolderProxy = proxyEvent._proxy;
});

``` 

## Direct transfer of utility tokens

```js
const txOptions = {
  from: relayer,
  gasPrice: gasPrice,
  gas: gas
};
const tokenRules = new OpenST.Helpers.TokenRules(tokenRulesAddress, web3Provider);
const receiver = '0xaabb1122....................'; 
const directTransferCallData = tokenRules.getDirectTransferExecutableData([receiver], [100]);
const nonce = 0; // Ephemeral key current nonce.
const tokenHolder = new OpenST.Helpers.TokenHolder(web3Provider, userTokenHolderProxy);
let transaction = {
  from: userTokenHolderProxy,
  to: tokenRulesAddress,
  data: directTransferCallData,
  nonce: nonce,
  callPrefix: tokenHolder.getTokenHolderExecuteRuleCallPrefix(),
  value: 0,
  gasPrice: 0,
  gas: 0
};
let ephemeralKeyAccountInstance; // Construct account instance of ephemeral key
const vrs = ephemeralKeyAccountInstance.signEIP1077Transaction(transaction);
tokenHolder.executeRule(tokenRulesAddress, directTransferCallData, nonce, vrs.r, vrs.s, vrs.v, txOptions).then(function(response){
  console.log("Successful transfer done!");
})

```

## Setup of PriceOracle 

Make sure PriceOracle contract is already deployed. Refer [PriceOracle](https://github.com/openst/ost-price-oracle) npm for setup of PriceOracle contract.

## Payment through PricerRule contract 

```js
async function pay() {
   let txOptions = {
      from: worker,
      gasPrice: gasPrice,
      gas: gas
   };
  const pricerRule = new OpenST.Helpers.Rules.PricerRule(web3Provider, pricerRuleAddress);
  let priceOracleAddress;
  const acceptanceMargin = '10000000000000000000';
  await pricerRule.addPriceOracle(priceOracleAddress, txOptions); // Set PriceOracle address in PricerRule.
  await pricerRule.setAcceptanceMargin('USD', acceptanceMargin, txOptions); // Set appropriate acceptance margin.
       
  txOptions = {
    from: relayer,
    gasPrice: gasPrice,
    gas: gas
  };
  const tokenHolder = new OpenST.Helpers.TokenHolder(web3Provider, userTokenHolderProxy);
  const receiver = '0xaabb1122....................'; 
  const nonce = 0; // Ephemeral key current nonce.
  let baseCurrencyIntendedPrice; // Current OST/USD price.
  const pricerRulePayCallData = pricerRule.getPayExecutableData(userTokenHolderProxy, [receiver], ['1000'], 'USD', baseCurrencyIntendedPrice);
  let transaction = {
    from: userTokenHolderProxy,
    to: pricerRuleAddress,
    data: pricerRulePayCallData,
    nonce: nonce,
    callPrefix: tokenHolder.getTokenHolderExecuteRuleCallPrefix(),
    value: 0,
    gasPrice: 0,
    gas: 0
  };
  let ephemeralKeyAccountInstance;
  const signatureObj = ephemeralKeyAccountInstance.signEIP1077Transaction(transaction);
  tokenHolder.executeRule(pricerRuleAddress, pricerRulePayCallData, nonce, signatureObj.r, signatureObj.s, signatureObj.v, txOptions).then( function() {
    console.log("Successful transfer done!");
  })
}
pay();

```

## Wallet Operations

### Add Wallet

```js
async function addWallet() {
    const txOptions = {
      from: relayer,
      gasPrice: gasPrice,
      gas: gas
    };
    const gnosisSafe = new Package.Helpers.GnosisSafe(gnosisSafeProxy, web3Provider);
    const threshold = 1;
    const ownerToAdd = '0xaabb1122....................';
    const ownerToAddWithThresholdCallData = gnosisSafe.getAddOwnerWithThresholdExecutableData(ownerToAdd, threshold);
    const nullAddress = '0x0000000000000000000000000000000000000000';
    const nonce = await gnosisSafe.getNonce();
    const safeTxData = await gnosisSafe.getSafeTxData(gnosisSafeProxy, 0, ownerToAddWithThresholdCallData, 0, 0, 0, 0, nullAddress, nullAddress, nonce);
    // 2. Generate EIP712 Signature.
    const signatureObj = await owner.signEIP712TypedData(safeTxData);
    await gnosisSafe.execTransaction(gnosisSafeProxy, 0, ownerToAddWithThresholdCallData, 0, 0, 0, 0, nullAddress, nullAddress, signatureObj.signature, txOptions);
};
addWallet();    

```

Similar above steps apply for removeWallet and replaceWallet operations.

### Authorize Session

```js
async function authorizeSession() {
    const txOptions = {
      from: relayer,
      gasPrice: gasPrice,
      gas: gas
    };
    const tokenHolder = new TokenHolder(web3Provider, userTokenHolderProxy);
    const sessionKeyToAuthorize = '0xaabb1122....................';
    const spendingLimit = '1000000000000';
    const expirationHeight = '100000000000';
    const nullAddress = '0x0000000000000000000000000000000000000000';
    const authorizeSessionCallData = tokenHolder.getAuthorizeSessionExecutableData(sessionKeyToAuthorize, spendingLimit, expirationHeight);
    const gnosisSafe = new Package.Helpers.GnosisSafe(gnosisSafeProxy, web3Provider);
    const nonce = await gnosisSafe.getNonce();
    const safeTxData = gnosisSafe.getSafeTxData(userTokenHolderProxy, 0, authorizeSessionCallData, 0, 0, 0, 0, nullAddress, nullAddress, nonce);
    const signatureObj = await owner.signEIP712TypedData(safeTxData);
    await gnosisSafe.execTransaction(userTokenHolderProxy, 0, authorizeSessionCallData, 0, 0, 0, 0, nullAddress, nullAddress, signatureObj.signature, txOptions);
};
authorizeSession(); 

```

Similar above steps apply for revokeSession and logout operations.

## ABI and BIN provider

openst.js comes with an abi-bin provider for managing abi(s) and bin(s).

The abiBinProvider provides abi(s) and bin(s) for the following contracts:

* [TokenHolder](https://github.com/openst/openst-contracts/blob/0.10.0-alpha.1/contracts/token/TokenHolder.sol) (TokenHolder contract deployed on UtilityChain)
* [TokenRules](https://github.com/openst/openst-contracts/blob/0.10.0-alpha.1/contracts/token/TokenRules.sol) (TokenRules contract deployed on UtilityChain)
* [PricerRule](https://github.com/openst/openst-contracts/blob/0.10.0-alpha.1/contracts/rules/PricerRule.sol) (PricerRule is a rule contract deployed on UtilityChain)
* [GnosisSafe](https://github.com/gnosis/safe-contracts/blob/v0.10.0/contracts/GnosisSafe.sol) (MultiSignature wallet with support for confirmations using signed messages)
* [DelayedRecoveryModule](https://github.com/openst/openst-contracts/blob/0.10.0-alpha.1/contracts/gnosis_safe_modules/DelayedRecoveryModule.sol) (Allows to replace an owner without Safe confirmations) 
* [CreateAndAddModules](https://github.com/gnosis/safe-contracts/blob/v0.10.0/contracts/libraries/CreateAndAddModules.sol) (Allows to create and add multiple module in one transaction)
* [UserWalletFactory](https://github.com/openst/openst-contracts/blob/0.10.0-alpha.1/contracts/proxies/UserWalletFactory.sol) (Creates proxy for GnosisSafe, TokenHolder and DelayedRecoveryModule)

```js

// Fetching ABI examples.
let abiBinProvider = new OpenST.AbiBinProvider();
const TokenHolderAbi = abiBinProvider.getABI('TokenHolder');

```

## Tests

Tests require docker-compose. To run the tests, execute below command from root directory.

```bash
    // For unit tests.
    npm run test
    // For integration tests.
    npm run test:integration

```   

CHANGELOG.md
# openst.js Change Log

## Version 0.10.0

* Deprecation of setup and helper files ([141](https://github.com/openst/openst.js/pull/141))
* ABIs and BINs are now accessed as a dependency on openst-contracts ([137](https://github.com/openst/openst.js/pull/137))
* Update PricerRule ABI/BIN with PricerRule:convertPayCurrencyToToken fix ([126](https://github.com/openst/openst.js/pull/126))
* Recovery contract interaction fixes ([#116](https://github.com/openst/openst.js/pull/116))
* PricerRule integration test and cleanup ([#114](https://github.com/openst/openst.js/pull/114))
* ProxyFactory address refactoring from createCompanyWallet method ([#113](https://github.com/openst/openst.js/pull/113)
* Integration with mosaic.js npm ([#109](https://github.com/openst/openst.js/pull/109))
* Docker Geth Integration: Using docker geth ([#108](https://github.com/openst/openst.js/pull/108))
* Delayed Recovery Module Setup ([#107](https://github.com/openst/openst.js/pull/107))
* Remove license from top of JS files ([#106](https://github.com/openst/openst.js/pull/106))
* MultiSig Operation: Logout all sessions ([104](https://github.com/openst/openst.js/pull/104))
* WalletOperations Test: Remove tokenRulesAddress while creating TokenHolder instance ([#91](https://github.com/openst/openst.js/pull/91))
* Remove pre commit hook from package.json which generates webpack ([#88](https://github.com/openst/openst.js/pull/88))
* Fix version to 0 in SignEIP1077Extension.js ([#87](https://github.com/openst/openst.js/pull/87))
* Wallet operations cleanup and refactoring ([#79](https://github.com/openst/openst.js/pull/79))
* Multisig operations : Change threshold ([#78](https://github.com/openst/openst.js/pull/78))
* Added multisig wallet operations ([#74](https://github.com/openst/openst.js/pull/74))
* Parse payCurrencyCode to bytes in PricerRule deployment ([#71](https://github.com/openst/openst.js/pull/71))
* User setup: Company to User Implementation ([#68](https://github.com/openst/openst.js/pull/68))
* Direct transfer of tokens using TokenRules contract ([#67](https://github.com/openst/openst.js/pull/67))
* Rule setup: Pricer rule setup and interaction ([#66](https://github.com/openst/openst.js/pull/66))
* Rename _createUserWallet to _createUserWalletRawTx ([#64](https://github.com/openst/openst.js/pull/64))
* Updated params _createUserWallet ([#61](https://github.com/openst/openst.js/pull/61))
* TokenRules : Rules registration framework for registration of rules getter to fetch registered rules ([#60](https://github.com/openst/openst.js/pull/60))
* Updated index.js to expose setup libraries ([#57](https://github.com/openst/openst.js/pull/57))
* User setup : Implement createUserWallet ([#55](https://github.com/openst/openst.js/pull/55))
* Interaction for MultiSig and TokenHolder master copy deployment ([#53](https://github.com/openst/openst.js/pull/53))
* Interaction for TokenRules deployment ([#51](https://github.com/openst/openst.js/issues/51))

lib/User.js
'use strict';

const AbiBinProvider = require('./AbiBinProvider');
const Contracts = require('./Contracts');
const Utils = require('../utils/Utils');

/**
 * Class is used to create a wallet of an user and configure it.
 */
class User {
  /**
   * @constructor
   *
   * @param {string} tokenHolderMasterCopy The address of a master copy of TokenHolder contract.
   * @param {string} gnosisSafeMasterCopy The address of a master copy of GnosisSafe contract.
   * @param {string} createAndAddModules The address of CreateAndAddModules contract.
   * @param {string} delayedRecoveryModuleMasterCopy The address of a master copy of
   *                                        DelayedRecoveryModule contract.
   * @param {string} eip20Token The address of EIP20Token contract of the economy.
   * @param {string} tokenRules The address of TokenRules contract.
   * @param {string} userWalletFactory The address of UserWalletFactory contract.
   * @param {string} proxyFactory The address of ProxyFactory contract.
   * @param {Web3} auxiliaryWeb3 Auxiliary chain's web3 object.
   */
  constructor(
    tokenHolderMasterCopy,
    gnosisSafeMasterCopy,
    delayedRecoveryModuleMasterCopy,
    createAndAddModules,
    eip20Token,
    tokenRules,
    userWalletFactory,
    proxyFactory,
    auxiliaryWeb3
  ) {
    this.tokenHolderMasterCopy = tokenHolderMasterCopy;
    this.gnosisSafeMasterCopy = gnosisSafeMasterCopy;
    this.delayedRecoveryModuleMasterCopy = delayedRecoveryModuleMasterCopy;
    this.createAndAddModules = createAndAddModules;
    this.eip20Token = eip20Token;
    this.tokenRules = tokenRules;
    this.userWalletFactory = userWalletFactory;
    this.proxyFactory = proxyFactory;
    this.auxiliaryWeb3 = auxiliaryWeb3;
    this.abiBinProvider = new AbiBinProvider();
  }

  /**
   * Generates DelayedRecoveryModule::setup() function's executable data.
   *
   * @param {string} recoveryOwnerAddress  An address that signs the "recovery
   *                              initiation/abortion" and "reset recovery owner"
   *                              requests.
   * @param {string} recoveryControllerAddress An address that relays signed requests of
   *                                  different types.
   * @param {string} recoveryBlockDelay A required number of blocks to pass to
   *                           be able to execute a recovery request.
   *
   * @returns {string} Executable data of the function.
   */
  getDelayedRecoveryModuleSetupData(recoveryOwnerAddress, recoveryControllerAddress, recoveryBlockDelay) {
    return this.auxiliaryWeb3.eth.abi.encodeFunctionCall(
      {
        name: 'setup',
        type: 'function',
        inputs: [
          {
            type: 'address',
            name: 'recoveryOwnerAddress'
          },
          {
            type: 'address',
            name: 'recoveryControllerAddress'
          },
          {
            type: 'uint256',
            name: 'recoveryBlockDelay'
          }
        ]
      },
      [recoveryOwnerAddress, recoveryControllerAddress, recoveryBlockDelay]
    );
  }

  /**
   * Generates ProxyFactory::createProxy() function's executable data to
   * create DelayedRecoveryModule proxy.
   *
   * @param {string} delayedRecoverySetupData Executable data of the DelayedRecoveryModule::setup()
   *                                  function.
   *
   * @returns {string} Executable data of the function.
   */
  getDelayedRecoveryModuleCreationData(delayedRecoverySetupData) {
    return this.auxiliaryWeb3.eth.abi.encodeFunctionCall(
      {
        name: 'createProxy',
        type: 'function',
        inputs: [
          {
            type: 'address',
            name: 'delayedRecoveryModuleMasterCopy'
          },
          {
            type: 'bytes',
            name: 'delayedRecoverySetupData'
          }
        ]
      },
      [this.delayedRecoveryModuleMasterCopy, delayedRecoverySetupData]
    );
  }

  /**
   * Generates CreateAndAddModules::createAndAddModules() function's executable
   * data to create and enable modules in GnosisSafe.
   *
   * @param {string} modulesCreationData An array of an executable data to create and enable
   *                                     modules within GnosisSafe during construction.
   *
   * @returns {string} Executable data of the function.
   */
  getCreateAndAddModulesData(modulesCreationData) {
    const ModuleDataWrapper = new this.auxiliaryWeb3.eth.Contract([
      {
        constant: false,
        inputs: [
          {
            name: 'data',
            type: 'bytes'
          }
        ],
        name: 'setup',
        type: 'function'
      }
    ]);

    // Remove method id (10) and position of data in payload (64)
    const reducedModulesCreationData = modulesCreationData.reduce(
      (acc, data) =>
        acc +
        ModuleDataWrapper.methods
          .setup(data)
          .encodeABI()
          .substr(74),
      '0x'
    );

    return this.auxiliaryWeb3.eth.abi.encodeFunctionCall(
      {
        name: 'createAndAddModules',
        type: 'function',
        inputs: [
          {
            type: 'address',
            name: 'proxyFactory'
          },
          {
            type: 'bytes',
            name: 'data'
          }
        ]
      },
      [this.proxyFactory, reducedModulesCreationData]
    );
  }

  /**
   * Generate the executable data for setup method of GnosisSafe contract.
   *
   * @param {Array} owners List of owners of the multisig contract for a user.
   * @param {string} threshold Number of required confirmations for a Safe transaction.
   * @param {string} ecoveryOwnerAddress  An address that signs the "recovery
   *                              initiation/abortion" and "reset recovery owner"
   *                              requests.
   * @param {string} recoveryControllerAddress An address that relays signed requests of
   *                                  different types.
   * @param {string} recoveryBlockDelay A required number of blocks to pass to
   *                           be able to execute a recovery request.
   *
   * @returns {string} GnosisSafe executable data.
   */
  getGnosisSafeData(owners, threshold, recoveryOwnerAddress, recoveryControllerAddress, recoveryBlockDelay) {
    const delayedRecoveryModuleSetupData = this.getDelayedRecoveryModuleSetupData(
      recoveryOwnerAddress,
      recoveryControllerAddress,
      recoveryBlockDelay
    );

    const delayedRecoveryModuleCreationData = this.getDelayedRecoveryModuleCreationData(delayedRecoveryModuleSetupData);

    const createAndAddModulesData = this.getCreateAndAddModulesData([delayedRecoveryModuleCreationData]);

    return this.auxiliaryWeb3.eth.abi.encodeFunctionCall(
      {
        name: 'setup',
        type: 'function',
        inputs: [
          {
            type: 'address[]',
            name: '_owners'
          },
          {
            type: 'uint256',
            name: '_threshold'
          },
          {
            type: 'address',
            name: 'to'
          },
          {
            type: 'bytes',
            name: 'data'
          }
        ]
      },
      [owners, threshold, this.createAndAddModules, createAndAddModulesData]
    );
  }

  /**
   * Returns TokenHolder setup executable data.
   *
   * @param {string} owner Owner address. Owner could be hardware wallet.
   * @param {Array} sessionKeys Array of session keys.
   * @param {Array} sessionKeysSpendingLimits Array of spending limits.
   * @param {Array} sessionKeysExpirationHeights Array of expiration heights.
   *
   * @returns {string} TokenHolder setup executable data.
   */
  getTokenHolderSetupExecutableData(owner, sessionKeys, sessionKeysSpendingLimits, sessionKeysExpirationHeights) {
    const contract = Contracts.getTokenHolder(this.auxiliaryWeb3, this.tokenHolderMasterCopy);
    return contract.methods
      .setup(
        this.eip20Token,
        this.tokenRules,
        owner,
        sessionKeys,
        sessionKeysSpendingLimits,
        sessionKeysExpirationHeights
      )
      .encodeABI();
  }

  /**
   * It is used for creation and configuration of gnosis safe and token holder
   * proxy contract for user.
   *
   * @param {Array} owners List of owners of the multisig.
   * @param {string} threshold Number of required confirmations for a Safe transaction.
   * @param {string} recoveryOwnerAddress  An address that signs the "recovery
   *                              initiation/abortion" and "reset recovery owner"
   *                              requests.
   * @param {string} recoveryControllerAddress An address that relays signed requests of
   *                                  different types.
   * @param {string} recoveryBlockDelay A required number of blocks to pass to
   *                           be able to execute a recovery request.
   * @param {Array} sessionKeys Session key addresses to authorize.
   * @param {Array} sessionKeysSpendingLimits Session key's spending limits.
   * @param {Array} sessionKeysExpirationHeights Session key's expiration heights.
   * @param {Object} txOptions Tx options.
   *
   * @returns {Promise<Object>} Promise that resolves to transaction receipt.
   */
  async createUserWallet(
    owners,
    threshold,
    recoveryOwnerAddress,
    recoveryControllerAddress,
    recoveryBlockDelay,
    sessionKeys,
    sessionKeysSpendingLimits,
    sessionKeysExpirationHeights,
    txOptions
  ) {
    const txObject = await this._createUserWalletRawTx(
      owners,
      threshold,
      recoveryOwnerAddress,
      recoveryControllerAddress,
      recoveryBlockDelay,
      sessionKeys,
      sessionKeysSpendingLimits,
      sessionKeysExpirationHeights
    );

    return Utils.sendTransaction(txObject, txOptions);
  }

  /**
   * Method for creation and configuration of token holder proxy contract for company
   * with hardware wallet as it's owner.
   *
   * @param {string} owner TokenHolder proxy owner address. It could be hardware wallet address.
   * @param {Array} sessionKeys Session key addresses to authorize.
   * @param {Array} sessionKeysSpendingLimits Session key's spending limits.
   * @param {Array} sessionKeysExpirationHeights Session key's expiration heights.
   *
   * @returns {Promise<Object>} Promise that resolves to Transaction receipt.
   */
  async createCompanyWallet(owner, sessionKeys, sessionKeysSpendingLimits, sessionKeysExpirationHeights, txOptions) {
    const txObject = await this._createCompanyWalletRawTx(
      owner,
      sessionKeys,
      sessionKeysSpendingLimits,
      sessionKeysExpirationHeights
    );

    return Utils.sendTransaction(txObject, txOptions);
  }

  /**
   * Private method used for creation and configuration of gnosis safe and
   * tokenholder contract for an user.
   *
   * @param {Array} owners List of owners of the multisig.
   * @param {string} threshold Number of required confirmations for a Safe transaction.
   * @param {string} recoveryOwnerAddress  An address that signs the "recovery
   *                              initiation/abortion" and "reset recovery owner"
   *                              requests.
   * @param {string} recoveryControllerAddress An address that relays signed requests of
   *                                  different types.
   * @param {string} recoveryBlockDelay A required number of blocks to pass to
   *                           be able to execute a recovery request.
   * @param {Array} sessionKeys Session key addresses to authorize.
   * @param {Array} sessionKeysSpendingLimits Session key's spending limits.
   * @param {Array} sessionKeysExpirationHeights Session key's expiration heights.
   *
   * @returns {Promise<Object>} Promise that resolves to raw transaction object.
   */
  _createUserWalletRawTx(
    owners,
    threshold,
    recoveryOwnerAddress,
    recoveryControllerAddress,
    recoveryBlockDelay,
    sessionKeys,
    sessionKeysSpendingLimits,
    sessionKeysExpirationHeights
  ) {
    const gnosisSafeData = this.getGnosisSafeData(
      owners,
      threshold,
      recoveryOwnerAddress,
      recoveryControllerAddress,
      recoveryBlockDelay
    );
    const contract = Contracts.getUserWalletFactory(this.auxiliaryWeb3, this.userWalletFactory);
    return Promise.resolve(
      contract.methods.createUserWallet(
        this.gnosisSafeMasterCopy,
        gnosisSafeData,
        this.tokenHolderMasterCopy,
        this.eip20Token,
        this.tokenRules,
        sessionKeys,
        sessionKeysSpendingLimits,
        sessionKeysExpirationHeights
      )
    );
  }

  /**
   * Private method used for creation and configuration of token holder proxy contract for company
   * with hardware wallet as it's owner.
   *
   * @param {string} owner TokenHolder proxy owner address. It could be hardware wallet address.
   * @param {Array} sessionKeys Session key addresses to authorize.
   * @param {Array} sessionKeysSpendingLimits Session key's spending limits.
   * @param {Array} sessionKeysExpirationHeights Session key's expiration heights.
   *
   * @returns {Promise<Object>} Promise that resolves to raw transaction object.
   */
  _createCompanyWalletRawTx(owner, sessionKeys, sessionKeysSpendingLimits, sessionKeysExpirationHeights) {
    const thSetupExecutableData = this.getTokenHolderSetupExecutableData(
      owner,
      sessionKeys,
      sessionKeysSpendingLimits,
      sessionKeysExpirationHeights
    );

    const contract = Contracts.getProxyFactory(this.auxiliaryWeb3, this.proxyFactory);
    return Promise.resolve(contract.methods.createProxy(this.tokenHolderMasterCopy, thSetupExecutableData));
  }
}

module.exports = User;
